temporal-sdk/src/main/java/io/temporal/worker/slotsupplier/SlotSupplier.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.slotsupplier;
+
+import java.util.Optional;
+
+public interface SlotSupplier<SlotInfo> {
+  // TODO: Needs to be cancellable
+  SlotPermit reserveSlot(SlotReservationContext ctx) throws InterruptedException;
+
+  Optional<SlotPermit> tryReserveSlot(SlotReservationContext ctx);
+
+  void markSlotUsed(SlotInfo info, SlotPermit permit);
+
+  void releaseSlot(SlotReleaseReason reason, SlotPermit permit);
```

Comment:
Can `releaseSlots` be blocking? If we had some sort of remote slot provider it seems like it could be?

Reply:
I don't think we can allow it to be blocking. The code has no expectation now that it might, and allowing it to could I think gum up stuff in unexpected ways. I mentioned in the proposal that users can delegate any blocking work to a background thread or to calls to reserve slot, and I think that's what they ought to do.